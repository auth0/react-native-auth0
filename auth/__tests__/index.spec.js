import Auth from '../';
import fetchMock from 'fetch-mock';

describe('auth', () => {

  const baseUrl = 'samples.auth0.com';
  const clientId = 'A_CLIENT_ID_OF_YOUR_ACCOUNT';
  const telemetry = {name: 'react-native-auth0', version: '1.0.0'};
  const redirectUri = 'https://mysite.com/callback';
  const state = 'a random state for auth';
  const tokens = {
    status: 200,
    body: {
      access_token: 'an access token',
      id_token: 'an id token',
      expires_in: 1234567890,
      state,
      scope: 'openid'
    },
    headers: { 'Content-Type': 'application/json' }
  };
  const oauthError = {
    status: 400,
    body: {
      error: 'invalid_request',
      error_description: 'Invalid grant'
    },
    headers: { 'Content-Type': 'application/json' }
  };
  const unexpectedError = {
    status: 500,
    body: 'Internal Server Error....',
    headers: { 'Content-Type': 'text/plain' }
  };
  const auth  = new Auth({baseUrl, clientId, telemetry});

  beforeEach(fetchMock.restore);

  describe('constructor', () => {
    it('should build with domain', () => {
      const auth = new Auth({baseUrl, clientId});
      expect(auth.clientId).toEqual(clientId);
    });

    it('should fail without clientId', () => {
      expect(() => new Auth({baseUrl})).toThrowErrorMatchingSnapshot();
    });

    it('should fail without domain', () => {
      expect(() => new Auth({clientId})).toThrowErrorMatchingSnapshot();
    });
  });


  describe('authorizeUrl', () => {
    it('should return default authorize url', () => {
      expect(auth.authorizeUrl({
        responseType: 'code',
        redirectUri,
        state: 'a_random_state'
      })).toMatchSnapshot();
    });

    it('should return default authorize url with extra parameters', () => {
      expect(auth.authorizeUrl({
        responseType: 'code',
        redirectUri,
        state: 'a_random_state',
        connection: 'facebook'
      })).toMatchSnapshot();
    });
  });

  describe('logoutUrl', () => {
    it('should return default logout url', () => {
      expect(auth.logoutUrl({})).toMatchSnapshot();
    });

    it('should return logout url with extra parameters', () => {
      expect(auth.logoutUrl({
        federated: true,
        clientId: 'CLIENT_ID',
        redirectTo: 'https://auth0.com'
      })).toMatchSnapshot();
    });

    it('should return logout url with skipping unknown parameters', () => {
      expect(auth.logoutUrl({
       federated: true,
        shouldNotBeThere: 'really'
      })).toMatchSnapshot();
    });
  });

  describe('code exchange', () => {
    it('should send correct payload', async () => {
      fetchMock.postOnce('https://samples.auth0.com/oauth/token', tokens);
      expect.assertions(1);
      await auth.exchange({code: 'a code', verifier: 'a verifier', redirectUri, state, scope: 'openid'});
      expect(fetchMock.lastCall()).toMatchSnapshot();
    });

    it('should return successful response', async () => {
      fetchMock.postOnce('https://samples.auth0.com/oauth/token', tokens);
      expect.assertions(1);
      const parameters = {code: 'a code', verifier: 'a verifier', redirectUri, state, scope: 'openid'};
      await expect(auth.exchange(parameters)).resolves.toMatchSnapshot();
    });

    it('should handle oauth error', async () => {
      fetchMock.postOnce('https://samples.auth0.com/oauth/token', oauthError);
      expect.assertions(1);
      const parameters = {code: 'a code', verifier: 'a verifier', redirectUri, state, scope: 'openid'};
      await expect(auth.exchange(parameters)).rejects.toMatchSnapshot();
    });

    it('should handle unexpected error', async () => {
      fetchMock.postOnce('https://samples.auth0.com/oauth/token', unexpectedError);
      expect.assertions(1);
      const parameters = {code: 'a code', verifier: 'a verifier', redirectUri, state, scope: 'openid'};
      await expect(auth.exchange(parameters)).rejects.toMatchSnapshot();
    });
  });

  describe('password realm', () => {
    const parameters = {username: 'info@auth0.com', password: 'secret pass', realm: 'Username-Password-Authentication', audience: 'http://myapi.com', scope: 'openid'};
    it('should send correct payload', async () => {
      fetchMock.postOnce('https://samples.auth0.com/oauth/token', tokens);
      expect.assertions(1);
      await auth.passwordRealm(parameters);
      expect(fetchMock.lastCall()).toMatchSnapshot();
    });

    it('should return successful response', async () => {
      fetchMock.postOnce('https://samples.auth0.com/oauth/token', tokens);
      expect.assertions(1);
      await expect(auth.passwordRealm(parameters)).resolves.toMatchSnapshot();
    });

    it('should handle oauth error', async () => {
      fetchMock.postOnce('https://samples.auth0.com/oauth/token', oauthError);
      expect.assertions(1);
      await expect(auth.passwordRealm(parameters)).rejects.toMatchSnapshot();
    });

    it('should handle unexpected error', async () => {
      fetchMock.postOnce('https://samples.auth0.com/oauth/token', unexpectedError);
      expect.assertions(1);
      await expect(auth.passwordRealm(parameters)).rejects.toMatchSnapshot();
    });
  });

  describe('refresh token', () => {
    const parameters = {refreshToken: 'a refresh token of a user', scope: 'openid'};
    it('should send correct payload', async () => {
      fetchMock.postOnce('https://samples.auth0.com/oauth/token', tokens);
      expect.assertions(1);
      await auth.refreshToken(parameters);
      expect(fetchMock.lastCall()).toMatchSnapshot();
    });

    it('should return successful response', async () => {
      fetchMock.postOnce('https://samples.auth0.com/oauth/token', tokens);
      expect.assertions(1);
      await expect(auth.refreshToken(parameters)).resolves.toMatchSnapshot();
    });

    it('should handle oauth error', async () => {
      fetchMock.postOnce('https://samples.auth0.com/oauth/token', oauthError);
      expect.assertions(1);
      await expect(auth.refreshToken(parameters)).rejects.toMatchSnapshot();
    });

    it('should handle unexpected error', async () => {
      fetchMock.postOnce('https://samples.auth0.com/oauth/token', unexpectedError);
      expect.assertions(1);
      await expect(auth.refreshToken(parameters)).rejects.toMatchSnapshot();
    });
  });

  describe('revoke token', () => {
    const parameters = {refreshToken: 'a refresh token of a user'};
    const success = {
      status: 200,
      body: null,
      headers: { 'Content-Type': 'application/json' }
    };
    it('should send correct payload', async () => {
      fetchMock.postOnce('https://samples.auth0.com/oauth/revoke', success);
      expect.assertions(1);
      await auth.revoke(parameters);
      expect(fetchMock.lastCall()).toMatchSnapshot();
    });

    it('should return successful response', async () => {
      fetchMock.postOnce('https://samples.auth0.com/oauth/revoke', success);
      expect.assertions(1);
      await expect(auth.revoke(parameters)).resolves.toMatchSnapshot();
    });

    it('should handle oauth error', async () => {
      fetchMock.postOnce('https://samples.auth0.com/oauth/revoke', oauthError);
      expect.assertions(1);
      await expect(auth.revoke(parameters)).rejects.toMatchSnapshot();
    });

    it('should handle unexpected error', async () => {
      fetchMock.postOnce('https://samples.auth0.com/oauth/revoke', unexpectedError);
      expect.assertions(1);
      await expect(auth.revoke(parameters)).rejects.toMatchSnapshot();
    });
  });

  describe('user info', () => {
    const parameters = {token: 'an access token of a user'};
    const success = {
      status: 200,
      body: {
        sub: '248289761001',
        name: 'Jane Doe',
        given_name: 'Jane',
        family_name: 'Doe',
        preferred_username: 'j.doe',
        email: 'janedoe@example.com',
        updated_at: 1497317424,
        picture: 'http://example.com/janedoe/me.jpg',
        'http://mysite.com/claims/customer': 192837465,
        'http://mysite.com/claims/status': 'closed'
      },
      headers: { 'Content-Type': 'application/json' }
    };
    it('should send correct payload', async () => {
      fetchMock.getOnce('https://samples.auth0.com/userinfo', success);
      expect.assertions(1);
      await auth.userInfo(parameters);
      expect(fetchMock.lastCall()).toMatchSnapshot();
    });

    it('should return successful oidc response', async () => {
      fetchMock.getOnce('https://samples.auth0.com/userinfo', success);
      expect.assertions(1);
      await expect(auth.userInfo(parameters)).resolves.toMatchSnapshot();
    });

    it('should return successful non-oidc response', async () => {
      fetchMock.getOnce('https://samples.auth0.com/userinfo', { sub: "auth0|1029837475" });
      expect.assertions(1);
      await expect(auth.userInfo(parameters)).resolves.toMatchSnapshot();
    });

    it('should handle oauth error', async () => {
      fetchMock.getOnce('https://samples.auth0.com/userinfo', oauthError);
      expect.assertions(1);
      await expect(auth.userInfo(parameters)).rejects.toMatchSnapshot();
    });

    it('should handle unexpected error', async () => {
      fetchMock.getOnce('https://samples.auth0.com/userinfo', unexpectedError);
      expect.assertions(1);
      await expect(auth.userInfo(parameters)).rejects.toMatchSnapshot();
    });
  });

  describe('change password', () => {
    const parameters = {email: 'info@auth0.com', connection: 'Username-Password-Authentication'};
    const success = {
      status: 200,
      body: "We've just sent you an email to reset your password.",
      headers: { 'Content-Type': 'text/html' }
    };
    it('should send correct payload', async () => {
      fetchMock.postOnce('https://samples.auth0.com/dbconnections/change_password', success);
      expect.assertions(1);
      await auth.resetPassword(parameters);
      expect(fetchMock.lastCall()).toMatchSnapshot();
    });

    it('should return successful response', async () => {
      fetchMock.postOnce('https://samples.auth0.com/dbconnections/change_password', success);
      expect.assertions(1);
      await expect(auth.resetPassword(parameters)).resolves.toMatchSnapshot();
    });

    it('should handle oauth error', async () => {
      fetchMock.postOnce('https://samples.auth0.com/dbconnections/change_password', oauthError);
      expect.assertions(1);
      await expect(auth.resetPassword(parameters)).rejects.toMatchSnapshot();
    });

    it('should handle unexpected error', async () => {
      fetchMock.postOnce('https://samples.auth0.com/dbconnections/change_password', unexpectedError);
      expect.assertions(1);
      await expect(auth.resetPassword(parameters)).rejects.toMatchSnapshot();
    });
  });

  describe('create user', () => {
    const parameters = {
      email: 'info@auth0.com',
      password: 'secret',
      connection: 'aconnection'
    };
    const success = {
      status: 200,
      body: {
        email: 'info@auth0.com',
        email_verified: false
      },
      headers: { 'Content-Type': 'application/json' }
    };
    const auth0Error = {
      status: 400,
      body: {
        "code": "user_exists",
        "description": "The user already exists.",
        "name": "BadRequestError",
        "statusCode": 400
      },
      headers: { 'Content-Type': 'application/json' }
    };

    it('should send correct payload', async () => {
      fetchMock.postOnce('https://samples.auth0.com/dbconnections/signup', success);
      expect.assertions(1);
      await auth.createUser(parameters);
      expect(fetchMock.lastCall()).toMatchSnapshot();
    });

    it('should send correct payload with username', async () => {
      fetchMock.postOnce('https://samples.auth0.com/dbconnections/signup', success);
      expect.assertions(1);
      await auth.createUser({ ...parameters, usename: 'info' });
      expect(fetchMock.lastCall()).toMatchSnapshot();
    });

    it('should send correct payload with metadata', async () => {
      fetchMock.postOnce('https://samples.auth0.com/dbconnections/signup', success);
      expect.assertions(1);
      await auth.createUser({ ...parameters, metadata: {customerId: 12345} });
      expect(fetchMock.lastCall()).toMatchSnapshot();
    });

    it('should return successful response', async () => {
      fetchMock.postOnce('https://samples.auth0.com/dbconnections/signup', success);
      expect.assertions(1);
      await expect(auth.createUser(parameters)).resolves.toMatchSnapshot();
    });

    it('should handle auth0 error', async () => {
      fetchMock.postOnce('https://samples.auth0.com/dbconnections/signup', auth0Error);
      expect.assertions(1);
      await expect(auth.createUser(parameters)).rejects.toMatchSnapshot();
    });

    it('should handle unexpected error', async () => {
      fetchMock.postOnce('https://samples.auth0.com/dbconnections/signup', unexpectedError);
      expect.assertions(1);
      await expect(auth.createUser(parameters)).rejects.toMatchSnapshot();
    });
  });
});