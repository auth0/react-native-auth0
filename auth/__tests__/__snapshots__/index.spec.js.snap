// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auth authorizeUrl should return default authorize url 1`] = `"https://samples.auth0.com/authorize?response_type=code&redirect_uri=https%3A%2F%2Fmysite.com%2Fcallback&state=a_random_state&client_id=A_CLIENT_ID_OF_YOUR_ACCOUNT&auth0Client=eyJuYW1lIjoicmVhY3QtbmF0aXZlLWF1dGgwIiwidmVyc2lvbiI6IjEuMC4wIn0%3D"`;

exports[`auth authorizeUrl should return default authorize url with extra parameters 1`] = `"https://samples.auth0.com/authorize?response_type=code&redirect_uri=https%3A%2F%2Fmysite.com%2Fcallback&state=a_random_state&connection=facebook&client_id=A_CLIENT_ID_OF_YOUR_ACCOUNT&auth0Client=eyJuYW1lIjoicmVhY3QtbmF0aXZlLWF1dGgwIiwidmVyc2lvbiI6IjEuMC4wIn0%3D"`;

exports[`auth change password should handle oauth error 1`] = `[invalid_request: Invalid grant]`;

exports[`auth change password should handle unexpected error 1`] = `[a0.response.invalid: Internal Server Error]`;

exports[`auth change password should return successful response 1`] = `Object {}`;

exports[`auth change password should send correct payload 1`] = `
Array [
  "https://samples.auth0.com/dbconnections/change_password",
  Object {
    "body": "{\\"email\\":\\"info@auth0.com\\",\\"connection\\":\\"Username-Password-Authentication\\",\\"client_id\\":\\"A_CLIENT_ID_OF_YOUR_ACCOUNT\\"}",
    "headers": Object {
      "Accept": "application/json",
      "Auth0-Client": "eyJuYW1lIjoicmVhY3QtbmF0aXZlLWF1dGgwIiwidmVyc2lvbiI6IjEuMC4wIn0=",
      "Content-Type": "application/json",
    },
    "method": "POST",
  },
]
`;

exports[`auth code exchange should handle oauth error 1`] = `[invalid_request: Invalid grant]`;

exports[`auth code exchange should handle unexpected error 1`] = `[a0.response.invalid: Internal Server Error]`;

exports[`auth code exchange should return successful response 1`] = `
Object {
  "accessToken": "an access token",
  "expiresIn": 1234567890,
  "idToken": "an id token",
  "scope": "openid",
  "state": "a random state for auth",
}
`;

exports[`auth code exchange should send correct payload 1`] = `
Array [
  "https://samples.auth0.com/oauth/token",
  Object {
    "body": "{\\"code\\":\\"a code\\",\\"code_verifier\\":\\"a verifier\\",\\"redirect_uri\\":\\"https://mysite.com/callback\\",\\"client_id\\":\\"A_CLIENT_ID_OF_YOUR_ACCOUNT\\",\\"grant_type\\":\\"authorization_code\\"}",
    "headers": Object {
      "Accept": "application/json",
      "Auth0-Client": "eyJuYW1lIjoicmVhY3QtbmF0aXZlLWF1dGgwIiwidmVyc2lvbiI6IjEuMC4wIn0=",
      "Content-Type": "application/json",
    },
    "method": "POST",
  },
]
`;

exports[`auth constructor should fail without clientId 1`] = `"Missing clientId in parameters"`;

exports[`auth constructor should fail without domain 1`] = `"Missing Auth0 domain"`;

exports[`auth create user should handle auth0 error 1`] = `[BadRequestError: The user already exists.]`;

exports[`auth create user should handle unexpected error 1`] = `[a0.response.invalid: Internal Server Error]`;

exports[`auth create user should return successful response 1`] = `
Object {
  "email": "info@auth0.com",
  "emailVerified": false,
}
`;

exports[`auth create user should send correct payload 1`] = `
Array [
  "https://samples.auth0.com/dbconnections/signup",
  Object {
    "body": "{\\"email\\":\\"info@auth0.com\\",\\"password\\":\\"secret\\",\\"connection\\":\\"aconnection\\",\\"client_id\\":\\"A_CLIENT_ID_OF_YOUR_ACCOUNT\\"}",
    "headers": Object {
      "Accept": "application/json",
      "Auth0-Client": "eyJuYW1lIjoicmVhY3QtbmF0aXZlLWF1dGgwIiwidmVyc2lvbiI6IjEuMC4wIn0=",
      "Content-Type": "application/json",
    },
    "method": "POST",
  },
]
`;

exports[`auth create user should send correct payload with metadata 1`] = `
Array [
  "https://samples.auth0.com/dbconnections/signup",
  Object {
    "body": "{\\"email\\":\\"info@auth0.com\\",\\"password\\":\\"secret\\",\\"connection\\":\\"aconnection\\",\\"user_metadata\\":{\\"customerId\\":12345},\\"client_id\\":\\"A_CLIENT_ID_OF_YOUR_ACCOUNT\\"}",
    "headers": Object {
      "Accept": "application/json",
      "Auth0-Client": "eyJuYW1lIjoicmVhY3QtbmF0aXZlLWF1dGgwIiwidmVyc2lvbiI6IjEuMC4wIn0=",
      "Content-Type": "application/json",
    },
    "method": "POST",
  },
]
`;

exports[`auth create user should send correct payload with username 1`] = `
Array [
  "https://samples.auth0.com/dbconnections/signup",
  Object {
    "body": "{\\"email\\":\\"info@auth0.com\\",\\"password\\":\\"secret\\",\\"connection\\":\\"aconnection\\",\\"client_id\\":\\"A_CLIENT_ID_OF_YOUR_ACCOUNT\\"}",
    "headers": Object {
      "Accept": "application/json",
      "Auth0-Client": "eyJuYW1lIjoicmVhY3QtbmF0aXZlLWF1dGgwIiwidmVyc2lvbiI6IjEuMC4wIn0=",
      "Content-Type": "application/json",
    },
    "method": "POST",
  },
]
`;

exports[`auth logoutUrl should return default logout url 1`] = `"https://samples.auth0.com/v2/logout"`;

exports[`auth logoutUrl should return logout url with extra parameters 1`] = `"https://samples.auth0.com/v2/logout?federated=true&client_id=CLIENT_ID"`;

exports[`auth logoutUrl should return logout url with skipping unknown parameters 1`] = `"https://samples.auth0.com/v2/logout?federated=true"`;

exports[`auth password realm should handle oauth error 1`] = `[invalid_request: Invalid grant]`;

exports[`auth password realm should handle unexpected error 1`] = `[a0.response.invalid: Internal Server Error]`;

exports[`auth password realm should return successful response 1`] = `
Object {
  "accessToken": "an access token",
  "expiresIn": 1234567890,
  "idToken": "an id token",
  "scope": "openid",
  "state": "a random state for auth",
}
`;

exports[`auth password realm should send correct payload 1`] = `
Array [
  "https://samples.auth0.com/oauth/token",
  Object {
    "body": "{\\"username\\":\\"info@auth0.com\\",\\"password\\":\\"secret pass\\",\\"realm\\":\\"Username-Password-Authentication\\",\\"audience\\":\\"http://myapi.com\\",\\"scope\\":\\"openid\\",\\"client_id\\":\\"A_CLIENT_ID_OF_YOUR_ACCOUNT\\",\\"grant_type\\":\\"http://auth0.com/oauth/grant-type/password-realm\\"}",
    "headers": Object {
      "Accept": "application/json",
      "Auth0-Client": "eyJuYW1lIjoicmVhY3QtbmF0aXZlLWF1dGgwIiwidmVyc2lvbiI6IjEuMC4wIn0=",
      "Content-Type": "application/json",
    },
    "method": "POST",
  },
]
`;

exports[`auth refresh token should handle oauth error 1`] = `[invalid_request: Invalid grant]`;

exports[`auth refresh token should handle unexpected error 1`] = `[a0.response.invalid: Internal Server Error]`;

exports[`auth refresh token should return successful response 1`] = `
Object {
  "accessToken": "an access token",
  "expiresIn": 1234567890,
  "idToken": "an id token",
  "scope": "openid",
  "state": "a random state for auth",
}
`;

exports[`auth refresh token should send correct payload 1`] = `
Array [
  "https://samples.auth0.com/oauth/token",
  Object {
    "body": "{\\"refresh_token\\":\\"a refresh token of a user\\",\\"scope\\":\\"openid\\",\\"client_id\\":\\"A_CLIENT_ID_OF_YOUR_ACCOUNT\\",\\"grant_type\\":\\"refresh_token\\"}",
    "headers": Object {
      "Accept": "application/json",
      "Auth0-Client": "eyJuYW1lIjoicmVhY3QtbmF0aXZlLWF1dGgwIiwidmVyc2lvbiI6IjEuMC4wIn0=",
      "Content-Type": "application/json",
    },
    "method": "POST",
  },
]
`;

exports[`auth revoke token should handle oauth error 1`] = `[invalid_request: Invalid grant]`;

exports[`auth revoke token should handle unexpected error 1`] = `[a0.response.invalid: Internal Server Error]`;

exports[`auth revoke token should return successful response 1`] = `Object {}`;

exports[`auth revoke token should send correct payload 1`] = `
Array [
  "https://samples.auth0.com/oauth/revoke",
  Object {
    "body": "{\\"token\\":\\"a refresh token of a user\\",\\"client_id\\":\\"A_CLIENT_ID_OF_YOUR_ACCOUNT\\"}",
    "headers": Object {
      "Accept": "application/json",
      "Auth0-Client": "eyJuYW1lIjoicmVhY3QtbmF0aXZlLWF1dGgwIiwidmVyc2lvbiI6IjEuMC4wIn0=",
      "Content-Type": "application/json",
    },
    "method": "POST",
  },
]
`;

exports[`auth user info should handle oauth error 1`] = `[invalid_request: Invalid grant]`;

exports[`auth user info should handle unexpected error 1`] = `[a0.response.invalid: Internal Server Error]`;

exports[`auth user info should return successful non-oidc response 1`] = `
Object {
  "sub": "auth0|1029837475",
}
`;

exports[`auth user info should return successful oidc response 1`] = `
Object {
  "email": "janedoe@example.com",
  "familyName": "Doe",
  "givenName": "Jane",
  "http://mysite.com/claims/customer": 192837465,
  "http://mysite.com/claims/status": "closed",
  "name": "Jane Doe",
  "picture": "http://example.com/janedoe/me.jpg",
  "preferredUsername": "j.doe",
  "sub": "248289761001",
  "updatedAt": 1497317424,
}
`;

exports[`auth user info should send correct payload 1`] = `
Array [
  "https://samples.auth0.com/userinfo",
  Object {
    "headers": Object {
      "Accept": "application/json",
      "Auth0-Client": "eyJuYW1lIjoicmVhY3QtbmF0aXZlLWF1dGgwIiwidmVyc2lvbiI6IjEuMC4wIn0=",
      "Authorization": "Bearer an access token of a user",
      "Content-Type": "application/json",
    },
    "method": "GET",
  },
]
`;
